File: lines.cpp
// lines.cpp
#include <iostream>
#include <vector>
#include <random>
#include <string>
#include <iomanip>
#include <ctime>
#include <map>

using namespace std;

struct Puzzle {
    int size;
    vector<pair<pair<int,int>, pair<int,int>>> pairs;

    Puzzle(int sz, unsigned seed) : size(sz) {
        mt19937 rng(seed);
        if (size == 6) {
            pairs = {
                {{0,0},{5,5}}, {{0,1},{4,3}}, {{1,0},{3,4}},
                {{2,2},{4,4}}, {{3,1},{5,3}}, {{0,5},{5,0}}
            };
        } else {
            uniform_int_distribution<int> dist(0, size-1);
            for (int i = 0; i < size/2; ++i) {
                auto a = make_pair(dist(rng), dist(rng));
                auto b = make_pair(dist(rng), dist(rng));
                pairs.push_back({a,b});
            }
        }
    }

    void display(bool showSolution) const {
        string letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        cout << "  ";
        for (int i = 0; i < size; ++i) cout << i+1 << " ";
        cout << endl;
        for (int r = 0; r < size; ++r) {
            cout << letters[r] << " ";
            for (int c = 0; c < size; ++c) {
                bool found = false;
                for (size_t idx = 0; idx < pairs.size(); ++idx) {
                    auto &p = pairs[idx];
                    if ((r == p.first.first && c == p.first.second) ||
                        (r == p.second.first && c == p.second.second)) {
                        cout << idx+1 << " ";
                        found = true;
                        break;
                    }
                }
                if (!found) cout << ". ";
            }
            cout << endl;
        }
    }
};

int main(int argc, char* argv[]) {
    int size = 6;
    bool showSolution = false;
    for (int i=1; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "--size" && i+1 < argc) size = stoi(argv[++i]);
        else if (arg == "--show-solution") showSolution = true;
    }
    unsigned seed = time(nullptr);
    Puzzle p(size, seed);
    p.display(showSolution);
    return 0;
}

File: lines.rs
// lines.rs
use clap::Parser;
use rand::Rng;
use std::collections::HashMap;

#[derive(Parser)]
#[command(name = "lines")]
struct Args {
    #[arg(short, long, default_value_t = 6)]
    size: usize,
    #[arg(long)]
    show_solution: bool,
    #[arg(long)]
    date: Option<String>,
}

struct Puzzle {
    size: usize,
    pairs: Vec<((usize, usize), (usize, usize))>,
}

impl Puzzle {
    fn new(size: usize, seed: u64) -> Self {
        let mut rng = rand::thread_rng();
        let mut pairs = Vec::new();
        if size == 6 {
            pairs = vec![
                ((0,0), (5,5)), ((0,1), (4,3)), ((1,0), (3,4)),
                ((2,2), (4,4)), ((3,1), (5,3)), ((0,5), (5,0)),
            ];
        } else {
            for _ in 0..size/2 {
                let a = (rng.gen_range(0..size), rng.gen_range(0..size));
                let b = (rng.gen_range(0..size), rng.gen_range(0..size));
                pairs.push((a, b));
            }
        }
        Puzzle { size, pairs }
    }

    fn display(&self, _show_solution: bool) {
        let letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        print!("  ");
        for i in 0..self.size {
            print!("{} ", i+1);
        }
        println!();
        for r in 0..self.size {
            print!("{} ", letters.chars().nth(r).unwrap());
            for c in 0..self.size {
                let mut found = false;
                for (idx, (a, b)) in self.pairs.iter().enumerate() {
                    if (r, c) == *a || (r, c) == *b {
                        print!("{} ", idx+1);
                        found = true;
                        break;
                    }
                }
                if !found {
                    print!(". ");
                }
            }
            println!();
        }
    }
}

fn main() {
    let args = Args::parse();
    let seed = match args.date {
        Some(d) => d.chars().fold(0, |acc, c| acc ^ (c as u64)),
        None => std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs(),
    };
    let puzzle = Puzzle::new(args.size, seed);
    puzzle.display(args.show_solution);
}
